jul-ecs-formatter/src/main/java/co/elastic/logging/jul/EcsFormatter.java

```diff
@@ -0,0 +1,103 @@
+/*-
+ * #%L
+ * Java ECS logging
+ * %%
+ * Copyright (C) 2019 - 2020 Elastic and contributors
+ * %%
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * #L%
+ */
+package co.elastic.logging.jul;
+
+import java.util.logging.Formatter;
+import java.util.logging.LogManager;
+import java.util.logging.LogRecord;
+
+import co.elastic.logging.EcsJsonSerializer;
+
+public class EcsFormatter extends Formatter {
+
+    private boolean stackTraceAsArray;
+    private String serviceName;
+    private boolean includeOrigin;
+    private String eventDataset;
+
+    /**
+     * Default constructor. Will read configuration from LogManager properties.
+     */
+    public EcsFormatter() {
+        serviceName = getProperty("co.elastic.logging.jul.EcsFormatter.serviceName", null);
+        includeOrigin = Boolean.getBoolean(getProperty("co.elastic.logging.jul.EcsFormatter.includeOrigin", "false"));
+        stackTraceAsArray = Boolean
+                .getBoolean(getProperty("co.elastic.logging.jul.EcsFormatter.stackTraceAsArray", "false"));
+        eventDataset = getProperty("co.elastic.logging.jul.EcsFormatter.eventDataset", null);
+        eventDataset = EcsJsonSerializer.computeEventDataset(eventDataset, serviceName);
+    }
+
+    @Override
+    public String format(final LogRecord record) {
+        final StringBuilder builder = new StringBuilder();
```

Comment:
```suggestion
        final StringBuilder builder = EcsJsonSerializer.getMessageStringBuilder();
```

Reply:
Not sure that this is the best idea.
Not sure that this is the best idea...
Looks like you try to cache string buffer (actually array of chars) and this means using ThreadLocal + `Arrays.fill(val, index, end, (byte)0);` call for each StringBuilder reset. Looks like allocated array may only grow and very fast came to Old Generation. At least or long living threads. IMO just let the GC doing its job. Modern JVM should have well optimized GC strategies for wide range of use cases.

... and some tests are failed after sitch to EcsJsonSerializer.getMessageStringBuilder() version. Looks likeresult strung is brocken,  not sure why. Will look at  this issue tomorrow.

Reply:
For JUL, there's probably no big benefit as it allocates lots of objects regardless. But in general, I think that libraries should try to minimize garbage so that the business logic has more GC resources.

> Modern JVM should have well optimized GC strategies for wide range of use cases.

Yes, that's mostly true. But even almost all modern GC algorithms do is that they cause unpredictable latency. If you're running a batch application it's not a big deal. If you're running an e-commerce shop, regular stop-the-world pauses can cause annoyingly large latencies for end-users. For high-frequency trading applications, a stop-the-world would be disastrous. 

This is a great article from log4j2 if you're interested: https://logging.apache.org/log4j/2.x/manual/garbagefree.html

But again, as JUL needs allocations anyways to create a string from the StringBuilder, let's leave it as-is 🙂.

Reply:
OK, article has sense. However still have doubts about  EcsJsonSerializer.getMessageStringBuilder(); in EcsFormatter. It was proposed to accumulate JSON result, but there are few issue with it:
1. Looks like EcsJsonSerializer.getMessageStringBuilder() used as "temp buffer" and and may be reused (including cleaned) few times from different places. E.g. see EcsJsonSerializer.formatThrowableAsArray() method. It cleans buffer few times (buffer.setLength(0); call). May be it's kind of "unexpected" use case, but this exectly why Jenkins CI is not happy now.
2. I used co.elastic.logging.log4j.EcsLayout as example, and its format() method also starts from `StringBuilder builder = new StringBuilder();` string. So, maybe autor of log4j.EcsLayout also meet issue from point #1 :)

For now I switched to `new StringBuffer()` approach and pretty sure that CI would be green. If we really need "garbage-free" behavior here, than we should rethink usage of "EcsJsonSerializer.getMessageStringBuilder()". One option may be keep separate buffer for EcsJsonSerializer.formatThrowableAsArray() method. And also we may try optimize co.elastic.logging.log4j.EcsLayout  in the same direction. However I think that it's out of scope for current PR and has sense to open separate request.